server/common/annotations.py

```diff
@@ -0,0 +1,231 @@
+from datetime import datetime
+import re
+from uuid import uuid4
+import os
+import pandas as pd
+from hashlib import blake2b
+import base64
+from server import __version__ as cellxgene_version
+import threading
+from server.common.errors import AnnotationsError, OntologyLoadFailure
+from server.common.data_locator import DataLocator
+from server.common.utils import series_to_schema
+import fsspec
+import fastobo
+import traceback  # use built-in formatter for SyntaxError
+from flask import session
+
+
+class Annotations(object):
+    """ baseclass for annotations, including ontologies"""
+
+    """ our default ontology is the PURL for the Cell Ontology.
+    See http://www.obofoundry.org/ontology/cl.html """
+    DefaultOnotology = "http://purl.obolibrary.org/obo/cl.obo"
+
+    def __init__(self):
+        self.ontology_data = None
+
+    def load_ontology(self, path):
+        """Load and parse ontologies - currently support OBO files only."""
+        if path is None:
+            path = self.DefaultOnotology
+
+        try:
+            with fsspec.open(path) as f:
+                obo = fastobo.iter(f)
+                terms = filter(lambda stanza: type(stanza) is fastobo.term.TermFrame, obo)
+                names = [tag.name for term in terms for tag in term if type(tag) is fastobo.term.NameClause]
+                self.ontology_data = names
+
+        except FileNotFoundError as e:
+            raise OntologyLoadFailure(f"Unable to find OBO ontology path: {path}") from e
+
+        except SyntaxError as e:
+            msg = ''.join(traceback.format_exception_only(SyntaxError, e))
+            raise OntologyLoadFailure(msg) from e
+
+        except Exception as e:
+            raise OntologyLoadFailure(f"Error loading OBO file {path}") from e
+
+    def get_schema(self, data_adaptor):
```

Comment:
What is the relationship between the data adaptor and the annotation object?  Are they mix-and-match?

I'm not sure that makes sense, eg, the call lower down to `data_adaptor.get_location()` makes a big assumption that the data adaptor even knows how annotations are stored.

Seems to me that they should be completely independent, allowing a mix-and-match composition.

Reply:
I replaced the logic here with a function on the adaptor called "get_last_mod_time".
We may want to rethink what gets presented in the annotations file.
I think it may be useful if the data_adaptors could have some meta data that had versioning information, rather than relying on file timestamps.  We could always fallback to  timestamps.